lib/AST/ShouldPrintForParseableInterface.cpp

```diff
@@ -0,0 +1,233 @@
+/// types are correct for non-resilient types.
+///
+/// For example:
+///
+/// ```
+/// private struct B {}
+/// public struct A {
+///   internal let b: B
+/// }
+/// ```
+/// Since `A`'s size depends on `B`, we need to print `B` even though it's
+/// private. `B`'s children will also be checked for types that contribute to
+/// its layout, and so on.
+class FindReferencedNonPublicTypeDecls: public ASTWalker {
+  using TypeDecls = SmallPtrSetImpl<TypeDecl *>;
+  SmallVector<ValueDecl *, 8> workList;
+
+  /// Finds all nominal types or typealiases referenced in the type being
+  /// walked.
+  class FindTypeDecls: public TypeWalker {
+    TypeDecls &foundDecls;
+    SmallVectorImpl<ValueDecl *> &workList;
+  public:
+    FindTypeDecls(TypeDecls &foundDecls, SmallVectorImpl<ValueDecl *> &workList)
+      : foundDecls(foundDecls), workList(workList) {}
+
+    void addIfNonPublic(TypeDecl *typeDecl) {
+      if (isPublicOrUsableFromInline(typeDecl))
+        return;
+      if (foundDecls.insert(typeDecl).second)
+        workList.push_back(typeDecl);
+    }
+
+    /// Finds all nominal types or typealiases referenced by the
+    Action walkToTypePre(Type type) override {
+      if (auto nominal =
+          dyn_cast<NominalOrBoundGenericNominalType>(type.getPointer()))
+        addIfNonPublic(nominal->getDecl());
+      if (auto enumType = dyn_cast<EnumType>(type.getPointer()))
+        addIfNonPublic(enumType->getDecl());
+      if (auto alias = dyn_cast<NameAliasType>(type.getPointer())) {
+        // Add the typealias to the found decls
+        auto decl = alias->getDecl();
+        addIfNonPublic(decl);
+
+        // Also walk into the RHS of the typealias to find extra types we need.
+        decl->getUnderlyingTypeLoc().getType().walk(*this);
+      }
+      return Action::Continue;
+    }
+  };
+  FindTypeDecls findTypeDecls;
+public:
+  FindReferencedNonPublicTypeDecls(TypeDecls &foundDecls):
+    findTypeDecls(foundDecls, workList) {}
+
+  void search(ModuleDecl *module) {
+    SmallVector<Decl *, 8> topLevelDecls;
+    module->getTopLevelDecls(topLevelDecls);
+    for (auto decl : topLevelDecls) {
+      auto vd = dyn_cast<ValueDecl>(decl);
+      if (!vd || !isPublicOrUsableFromInline(vd))
+        continue;
+      workList.push_back(vd);
+    }
+
+    while (!workList.empty()) {
+      Decl *d = workList.pop_back_val();
+      d->walk(*this);
+    }
+  }
+
+  void visitAbstractStorageDecl(AbstractStorageDecl *decl) {
+    if (!contributesToParentTypeStorage(decl))
+      return;
+    decl->getInterfaceType().walk(findTypeDecls);
+  }
+
+  void visitEnumElementDecl(EnumElementDecl *decl) {
+    auto params = decl->getParameterList();
+    if (!params) return;
+    for (auto param : *params) {
+      param->getInterfaceType().walk(findTypeDecls);
+    }
+  }
+
+  void visitNominalTypeDecl(NominalTypeDecl *decl) {
+    for (auto inherited : decl->getInherited())
+      inherited.getType().walk(findTypeDecls);
```

Comment:
Uh, hm. If we don't print extensions, we might not get all the protocol conformances, which means we might not satisfy a generic constraint where this type is used. Is that a problem with our plan?

Reply:
Oh, yes that is a problem. Hmm...should I add extensions of this type to the printed set as well?

Reply:
That might work for now. We probably do need to sit down and think about whether printing some kind of "layout token" would be easier all around, though.

Reply:
Or rather, walk declared conformances in extensions of this type, marking those for inclusion?